presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,103 @@ private PinotExpression handleLogicalBinary(
+        right = handleTimeStampAndDateCast(arguments.get(0), right);
+        return derived(format("(%s %s %s)", left, operator, right));
+    }
+
+    private boolean isDateOrTimestampExpression(List<RowExpression> arguments)
+    {
+        for (RowExpression rowExpression : arguments) {
+            if (!(rowExpression instanceof CallExpression)) {
+                continue;
+            }
+            CallExpression callExpr = (CallExpression) rowExpression;
+            if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
+                continue;
+            }
+            if (callExpr.getArguments().size() != 1) {
+                continue;
+            }
+            Type inputType = callExpr.getArguments().get(0).getType();
+            Type expectedType = callExpr.getType();
+            if (inputType == DateType.DATE && (expectedType == TimestampType.TIMESTAMP || expectedType == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE)) {
+                return true;
+            }
+            if ((inputType == TimestampType.TIMESTAMP || inputType == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE) && expectedType == DateType.DATE) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private String handleTimeStampAndDateCast(RowExpression rowExpression, String timeStampOrDateString)
+    {
+        if (!(rowExpression instanceof CallExpression)) {
+            return timeStampOrDateString;
+        }
+        CallExpression callExpr = (CallExpression) rowExpression;
+        if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
+            return timeStampOrDateString;
+        }
+        if (callExpr.getArguments().size() != 1) {
+            return timeStampOrDateString;
+        }
+
+        // Handle the binary comparison logic of <DATE/TIMESTAMP field> <binary ops> <DATE/TIMESTAMP literal>.
+        // Pinot underlying stores time as:
+        //   - `DATE`: Stored as `INT`/`LONG` `daysSinceEpoch` value
+        //   - `TIMESTAMP`: Stored as `LONG` `millisSinceEpoch` value.
+        // In order to push down this predicate, we need to convert the literal to the type of Pinot time field.
+        // Below code compares the time type of both side:
+        //   - if same, then directly push down.
+        //   - if not same, then convert the literal time type to the field time type.
+        // Take an example of comparing a `DATE` field to a `TIMESTAMP` literal:
+        //   - Sample predicate: `WHERE eventDate < current_time`.
+        //   - Input type is the `eventDate` field data type, which is `DATE`.
+        //   - Expect type is the right side `literal type`, which means right side is `TIMESTAMP`.
+        // So below code will convert `current_time` in `millisSinceEpoch` value to `daysSinceEpoch` value.
+        // Then the comparison underlying inside Pinot will work.
```

Comment:
This will enable to work inside Pinot. 

I think the use of the word underlying is superfluous.